File: collection/All/untitled/src/BinaryTrees/MinimumValueInTree.java
package BinaryTrees;

public class MinimumValueInTree {

    private Node root = null;

    public void calculateMinimumValue() {
        Node current = root;
        int result = min(current);
        System.out.println("Result:->: "+result);
    }

    private int min(Node root) {

        if(root==null){
            return 0;
        }

        // Root Node
        int res = root.value;
        // Left Node
        int leftNode = min(root.left);
        // Right Node
        int rightNode = min(root.right);


        /**
         * LOGIC: Assume root value is the largest value
         *  (1) Compare root with left --> If left is larger value then store left in root
         *  (2) Compare root with right --> If right is larger value then store right in root
         */

        if (leftNode > res){
            res = leftNode;
        }

        if (rightNode > res){
            res = rightNode;
        }

        return res;
    }


    /** ************************ HELPER FUNCTIONS ********************** **/
    /**
     * Inserting a Node
     */
    public void insert(int value) {

        Node current = root;

        if(current==null){
            // There are no elements inserted, so we perform the first insertion operation
            root = createNode(value);
        }else{
            Node nodeToInsert = createNode(value);
            // Iterate to the leaf node from the root where we need to perform insertion
            while (current!=null){
                // Either we insert to left of root or we insert to right of root
                if(value < current.value){
                    if(current.right==null){
                        current.left = nodeToInsert;
                        break;
                    }else{
                        current = current.left;
                    }
                }else{
                    if(current.right==null){
                        current.right = nodeToInsert;
                        break;
                    }else{
                        current = current.right;
                    }
                }
            }
        }

    }

    public Node printResult(){
        return root;
    }

    private Node createNode(int value) {
        return new Node(value);
    }

    private class Node {
        Node left;
        Node right;
        int value;
        public  Node(int value){
            this.value = value;
        }

        @Override
        public String toString() {
            return "Node=" + value;
        }
    }
    /** ************************ HELPER FUNCTIONS ********************** **/
}

File: collection/Trees/BinaryTree/BinarySearchTree/Programs/MinimumValueInTree/MinimumValueInTree.kt
class MinimumValueInTree {

    // Root Node
    var root : Node? = null
    // Keeping track of size
    private var size = 0


    class Node(val info : Int) {
        var leftNode : Node? = null
        var rightNode : Node? = null
        override fun toString() = "Node -> $info"
    }

    /**
     * Printing the root node
     */
    fun printNode() = root

    /**
     * Creating a new node
     */
    fun createNode(info : Int) : Node = Node(info)

    /**
     * Inserting a new element
     */
    fun insert(info : Int) {

        val newNode = Node(info)

        if(size==0){
            // If the element we are inserting is the first element, Then it becomes the root node
            root = newNode
            size+=1
        }else{
            // At-least there is one element ----> Now determine if we can add to left or if we can add to right
            var current = root

            while(current!=null){

                if(current.info==info){
                    // Don't do anything
                }else if(info<current.info){
                    if(current.leftNode==null){
                        // We reached to leaf position
                        // Insert to the left of the tree
                        current.leftNode = newNode
                        size += 1
                        break
                    }else{
                        // Traverse to left of the current node
                        current = current.leftNode
                    }
                }else{
                    if(current.rightNode==null){
                        // We reached to leaf position
                        // Insert to the right of the tree
                        current.rightNode = newNode
                        size += 1
                        break
                    }else{
                        // Traverse to right of the current node
                        current = current.rightNode
                    }
                }

            }
        }



    }


    fun findMinimum(): Int? {
        if(size==0){
            return -1;
        }else{
            if (size==1){
                return root?.info
            }

            var current = root;
            while (current?.leftNode!=null) {
                current = current.leftNode
            }
            return current?.info
        }
    }


}